slatedb/src/clock.rs

```diff
@@ -58,27 +58,56 @@ pub trait SystemClock: Debug + Send + Sync {
 /// specified duration has elapsed. This is to mimic Tokio's ticker behavior.
 pub struct SystemClockTicker<'a> {
     clock: &'a dyn SystemClock,
-    duration: Duration,
-    first_tick: bool,
+    duration: TimeDelta,
+    last_tick: AtomicI64,
 }
 
 impl<'a> SystemClockTicker<'a> {
     fn new(clock: &'a dyn SystemClock, duration: Duration) -> Self {
+        let duration = TimeDelta::from_std(duration).expect("duration is out of range");
         Self {
             clock,
             duration,
-            first_tick: true,
+            last_tick: AtomicI64::new(i64::MIN),
         }
     }
 
-    pub fn tick(&mut self) -> Pin<Box<dyn Future<Output = ()> + Send + 'a>> {
-        // Tokio's ticker ticks immediately when the first tick() is called.
-        // Let's emulate that behavior in our ticker.
-        if self.first_tick {
-            self.first_tick = false;
-            Box::pin(async {})
+    /// Returns a future that emits a signal every `duration` interval. The next tick is
+    /// calculated as last_tick + duration. The first tick will complete immediately.
+    /// This is to mimic Tokio's ticker behavior.
+    ///
+    /// If the clock advances more than the duration between `tick()` calls, the ticker
+    /// will tick immediately.
+    pub fn tick(&self) -> Pin<Box<dyn Future<Output = ()> + Send + '_>> {
```

Comment:
why change this from `&mut`? If its `&mut` you can just use a regular i64 rather than an atomic one. Not sure what it would really mean to have multiple concurrent calls to tick anyway

Reply:
Ah, I think I just go carried away with the fact that I _could_ do it. 😅 But you're right; my intent was to keep this as close to the Tokio ticker, which is also mut. And there are few (no?) use cases for concurrent ticks.